restricting file upload types

kkarpieszuk opened the pull request:
Fixes #35 

- adds file types natively supported by WP to the Settings > File upload types table so user can now disable them if needed

Test steps:
- activate fut plugin
- go to wp-admin > settings > file upload types and check if file types with description "WordPress natively registered type" are added and automatically enabled
- disable some poplar native file type (for example .jpeg) and save
- install wpforms and create form with file upload field
- preview form and try to upload .jpeg file -> you should see error file extension is not supported
- go to  wp-admin > settings > file upload types and re-enable .jpeg file extension
- go to form and try to upload .jpeg file -> you should be able to do this

cheh reviewed (changes requested):
@kkarpieszuk sorry for being silent for so long time 🙏 and thanks for handling it by yourself!

I've made a first round of reviews and here is my feedback:

1. There is an issue once I've checked out to PR branch without additional actions - *.jpg file type is not allowed, [see](https://a.supportally.com/TFeHTF). The same with other natively registered types, e.g. https://a.supportally.com/4NWVC2, https://a.supportally.com/TUH50E. Env.: WP 5.2 / PHP 5.6.
2. See comment below.

kkarpieszuk commented:
@cheh I have fixed issue with missing option to upload native type even though it was checked as available

additionally, I have applied all your suggested changes. In a few comments you asked why I decided for some approach and my general answer is: it was because I changed strategy during the work on this and sometimes I forgot to clean the older apporach (like injecting Plugin class into Allowed). I cleared all such cases and now code should be simpler.

cheh commented:
Unfortunately, there is `PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /plugins/file-upload-types/src/functions.php on line 31`. TBH, I don't have any idea yet why it happens.
One thing which I need to mention - I've caught in on PHP 5.6 / WP 5.2.

kkarpieszuk commented:
I moved it to separate line for better clarity (also in one more place where I wrapped it in brackets in the same way). @cheh please continue

cheh commented on src/Migrations/Dispatcher.php:
```diff
@@ -0,0 +1,69 @@
+		$option_changed = false;
+
+		foreach ( $all_migrations as $name => $callback ) {
+			if ( ! isset( $already_run[ $name ] ) && is_callable( $callback ) && $callback() ) {
```
Please, add a blank line before `if`.

cheh commented on src/Migrations/Dispatcher.php:
```diff
@@ -0,0 +1,69 @@
+		foreach ( $all_migrations as $name => $callback ) {
+			if ( ! isset( $already_run[ $name ] ) && is_callable( $callback ) && $callback() ) {
+					$already_run[ $name ] = 1;
+					$option_changed       = true;
```
Please, remove one extra tab-indent here and above.

cheh commented on src/Migrations/Dispatcher.php:
```diff
@@ -0,0 +1,69 @@
+		$all_migrations = $this->get_migrations_list();
+		$option_changed = false;
+
+		foreach ( $all_migrations as $name => $callback ) {
```
`$this->get_migrations_list()` could be used instead of `$all_migrations`.

cheh commented on src/Settings.php:
```diff
@@ -290,6 +292,10 @@ public function display_types_table() { // phpcs:ignore Generic.Metrics.NestingL
 				endif;
+
+					// phpcs:disable WPForms.Comments.PHPDocHooks.RequiredHookDocumentation
+					do_action( 'fileuploadtypes_settings_display_types_table_after_enabled_types' );
```
`awesomemotive/wpforms-phpcs` package has an issue and notify about an invalid hook name incorrectly. For better readability the `FileUploadTypes` namespace should be converted to `file_upload_types`. 

Could you please change nabes for all new hooks?
Obviously, you will have to use `phpcs:disable` / `phpcs:enable` for reducing a noise.

kkarpieszuk replied:
done

cheh commented on src/Allowed.php:
```diff
@@ -0,0 +1,131 @@
+	 *
+	 * @var array
+	 */
+	private $enabled_types;
```
It could be initialized as an empty array.

cheh reviewed (changes requested):
There is an interesting case when the user adds a fake MIME Type with an extension that already exists - https://a.supportally.com/iTIwIg. Personally, I expected that I wasn't able to upload `*.numbers` file.

kkarpieszuk commented on src/Restrict/Native/Admin.php:
```diff
@@ -0,0 +1,214 @@
+		add_filter( 'fileuploadtypes_settings_transform_native_type', [ $this, 'get_upload_type' ] );
+	}
+
+	// phpcs:disable Generic.Metrics.CyclomaticComplexity.TooHigh
```
it is needed in fact: without this, phpcs says the complexity of this method is too high.

kkarpieszuk commented on src/Restrict/Native/Admin.php:
```diff
@@ -0,0 +1,214 @@
+
+			remove_filter( 'upload_mimes', [ $this->allowed, 'allowed_types' ] );
+
+			self::$unfiltered_types = get_allowed_mime_types();
```
"native" name is a bit misinforming here, but I didn't find anything better. In fact the logic here is to add support for any file types which are not coming from our json file. 

So it doesn't matter if the file type is natively added by WP plugin or some plugin, other than File Upload Types plugin. 

Therefore, the logic here is correct: we just want to get all file types which are not affected by FUT plugin. We should keep this piece of code as is.

kkarpieszuk commented on src/Restrict/Native/Admin.php:
```diff
@@ -0,0 +1,214 @@
+		if ( ! empty( $native_types ) ) :
+
+		?>
+		<tr class="section">
```
I can completely do this, but I vote to create an GH issue for this and handle separately as this is not 100% related to the task. I mean this piece of code is, but what do you think about introducing templates for all html we already have? This code to echo pieces of `<table>` is here and there: it would be great to convert them to templates as a separate task, what to you say?

kkarpieszuk commented on src/Migrations/Dispatcher.php:
```diff
@@ -0,0 +1,69 @@
+	 *
+	 * @since {VERSION}
+	 *
+	 * @return void
```
changed

kkarpieszuk commented on src/Migrations/Dispatcher.php:
```diff
@@ -0,0 +1,69 @@
+	 *
+	 * @return void
+	 */
+	public function run_migrations() {
```
changed

kkarpieszuk commented:
@cheh I've applied most of your suggestions and if I didn't I provided my reasoning comments why not.

about this:

> There is an interesting case when the user adds a fake MIME Type with an extension that already exists - https://a.supportally.com/iTIwIg. Personally, I expected that I wasn't able to upload *.numbers file.

Yup, but the same is in `develop` branch so it is not related to the task. 

In general I found - but I didn't explore it deeply yet, I am still quite new to FUT plugin and mime types handling in WP - WordPress probably in fact ignores mime types when restricting and looks only on file extension. Weird and I might be wrong, but it requires deeper  investigation and that's why I suggest a separate ticket.

So to sum up again:
- I've applied suggested changes when neccessary
- I've provided comments why I can't change some parts
- I suggest two separate tickets: one for html templates and one for the mime issue
